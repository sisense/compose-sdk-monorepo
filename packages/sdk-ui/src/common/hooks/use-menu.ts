import { TranslatableError } from '@/translation/translatable-error.js';
import { useMenuContext } from '../components/menu/menu-context.js';

/**
 * Hook to access menu API.
 *
 * Throws an error if used outside the menu context.
 *
 * @internal
 */
export const useMenu = () => {
  const menuSettings = useMenuContext();

  if (!menuSettings) {
    throw new TranslatableError('errors.missingMenuRoot');
  }

  const { openMenu, closeMenu } = menuSettings;

  return {
    openMenu,
    closeMenu,
  };
};
