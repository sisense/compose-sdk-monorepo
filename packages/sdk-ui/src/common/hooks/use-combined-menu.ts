import { useCallback, useRef } from 'react';
import { MenuOptions } from '../components/menu/types.js';
import { useMenu } from './use-menu.js';

export type CombineMenusFn = (menusOptions: MenuOptions[]) => MenuOptions;

export interface CombinedMenuParams {
  /** Function to combine the current menu options with the captured menu options */
  combineMenus: CombineMenusFn;
}

/**
 * Custom hook to manage a combined menu flow,
 * allowing to combine multiple menus options opened one after the other.
 *
 * @internal
 */
export const useCombinedMenu = ({ combineMenus }: CombinedMenuParams) => {
  const capturedMenusOptions = useRef<MenuOptions[]>([]);
  const menuApi = useMenu();

  const openCombinedMenu = useCallback(() => {
    const combinedMenusOptions =
      capturedMenusOptions.current.length === 1
        ? capturedMenusOptions.current[0]
        : combineMenus(capturedMenusOptions.current);

    menuApi.openMenu({
      ...combinedMenusOptions,
      onClose: () => {
        capturedMenusOptions.current = [];
      },
    });
  }, [menuApi, combineMenus]);

  const openMenu = useCallback(
    (options: MenuOptions) => {
      capturedMenusOptions.current.push(options);

      openCombinedMenu();
    },
    [openCombinedMenu],
  );

  return {
    openMenu,
  };
};
