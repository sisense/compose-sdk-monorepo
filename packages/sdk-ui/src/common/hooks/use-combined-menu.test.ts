/* eslint-disable vitest/no-commented-out-tests */
import { renderHook } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, Mock } from 'vitest';
import { useCombinedMenu, IsTargetMenuFn, CombineMenusFn } from './use-combined-menu';
import { useMenu } from './use-menu';

// Mock the useMenu hook so we can spy on its return value
vi.mock('./use-menu', () => ({
  useMenu: vi.fn(),
}));

describe('useCombinedMenu', () => {
  const openMenuMock = vi.fn();

  // Each test re-mocks and resets function calls
  beforeEach(() => {
    vi.clearAllMocks();
    (useMenu as unknown as Mock).mockReturnValue({
      openMenu: openMenuMock,
    });
  });

  it('should open the menu with currentMenuOptions when openMenu is called', () => {
    const isTargetMenu: IsTargetMenuFn = () => false;
    const combineMenus: CombineMenusFn = (current) => current;

    const { result } = renderHook(() => useCombinedMenu({ isTargetMenu, combineMenus }));

    result.current.openMenu({
      position: {
        left: 42,
        top: 69,
      },
      itemSections: [{ sectionTitle: 'Some Section', items: [] }],
    });

    // It should have been called with our menu options
    expect(openMenuMock).toHaveBeenCalledOnce();
    expect(openMenuMock.mock.calls[0]).toMatchSnapshot();
  });
});
