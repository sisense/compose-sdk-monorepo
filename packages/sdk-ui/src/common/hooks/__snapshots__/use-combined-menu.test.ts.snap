// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useCombinedMenu > should open the menu with currentMenuOptions when openMenu is called 1`] = `
[
  {
    "itemSections": [
      {
        "items": [],
        "sectionTitle": "Some Section",
      },
    ],
    "onClose": [Function],
    "position": {
      "left": 42,
      "top": 69,
    },
  },
]
`;
